### ISSUE TITLE
WSL support for azcopy

### ISSUE THREAD

**Issue by joshuafernandes** [Azure/azure-storage-azcopy#2646]
##### Note: The version is visible when running AzCopy without any argument
azcopy version 10.24.0
### Which platform are you using? (ex: Windows, Mac, Linux)
Windows 11 with WSL (ubuntu 22.04)

##### Note: Please remove the SAS to avoid exposing your credentials. If you cannot remember the exact command, please retrieve it from the beginning of the log file.
export AZCOPY_AUTO_LOGIN_TYPE=DEVICE
 azcopy list "https://<nameofaccount>.blob.core.windows.net"
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code P.....  to authenticate.

INFO: Logging in under the "Common" tenant. This will log the account in under its home tenant.
INFO: If you plan to use AzCopy with a B2B account (where the account's home tenant is separate from the tenant of the target storage account), please sign in under the target tenant with --tenant-id
INFO: Login succeeded.
INFO: Authenticating to source using Azure AD

### What problem was encountered?
Azcopy doesn't work at all on wsl, have attemped the autologin as per https://github.com/Azure/azure-storage-azcopy/wiki/Improved-login-support-for-AzCopy-commands-(with-in-memory-secret-store). Have attempted the `keyctl` workaround. It gets past the login and just stalls

### How can we reproduce the problem in the simplest way?
1. Win11 with WSL
2. Install azcopy
3. Attempt to list a storageaccount, or write to it 

### Have you found a mitigation/solution?
Nope, none appear to exist

**Comment by siminsavani-msft** [comment]
Hi @joshuafernandes ! I was able to reproduce your issue. Could you please try using SAS to complete your job while we investigate this further?

**Comment by dironco-industries** [comment]
> Hi @joshuafernandes ! I was able to reproduce your issue. Could you please try using SAS to complete your job while we investigate this further?

I expand this problem,
It's not just from WSL in Win11, azcopy using "Microsoft Entra" login, it doesn't work from any Linux machine. You should always use SAS.

**Comment by gabe-microsoft** [comment]
This is unfortunately a very old bug. See #1922 and #2274. 

However, I recently discovered a workaround if you use the Azure CLI as well.

```bash
# Configure AzCopy to auto-login using Azure CLI.
export AZCOPY_AUTO_LOGIN_TYPE=AZCLI

# Login using Azure CLI (if not logged in already).
az login

# Run any AzCopy command.
azcopy copy https://mysa.blob.core.windows.net/mycontainer/myblob .
```

**Comment by smalis-msft** [comment]
I believe this broke with azcopy v10.22.0. Using v10.21.2 still works for my scenarios.

**Comment by andyneff** [comment]
I did a little digging into what is and isn't working:

## TL;DR:

`azcopy` cannot read nor write from the `_uid_ses.1000` WSL kering session, but it can with other sessions.

## Background

All of the WSL distros share a common keyring, (I'm guessing inherited from `/init`/kernel magic). It looks like this:

```
$ keyctl show
Session Keyring
 *** --alswrv   1000 65534  keyring: _uid_ses.1000
 *** --alswrv   1000 65534   \_ keyring: _uid.1000
 *** --alswrv   1000  1000       \_ keyring: _krb
 *** --alswrv   1000  1000           \_ user: krb_ccache:primary
```

where `1000` is the default UID for WSL. Opening a completely different distro (e.g. `fedoraremix` vs `Ubuntu-24.04`) actually have the same instance of this `_uid_ses.1000` keyring. I can even create a key in one OS and see it in the other. These are in fact a shared keyring session (cool!).

Running `keyctl session my_new_session` in both distro will result in non-shared keyrings sessions, but running `keyctl session _uid_ses.1000` will go back to the shared keyring.

## What isn't working?

When you run a separate session in WSL, (i.e. `keyctl session azcopy_session`), you `azcopy` is able to read and write to the keyring.

After the first login, I see:

```
$ keyctl show
Session Keyring
 *** --alswrv   1000  1000  keyring: azcopy_session
 *** --alswrv   1000  1000   \_ user: AzCopyOAuthTokenCache
```

So in `azcopy_session`, `azcopy` is able to read and write the `AzCopyOAuthTokenCache`, and everything works. I went ahead and took the `AzCopyOAuthTokenCache` key and copied it to `_uid_ses.1000` , and that did not work either (but it did for `test_session_2`). This informs me that not only is `azcopy` unable to write to `_uid_ses.1000`, but it cannot read from `_uid_ses.1000` either.

## Is v10.21.2 working?

TL;DR no

I tried [10.21.2](https://azcopyvnext.azureedge.net/releases/release-10.21.2-20231106/azcopy_linux_amd64_10.21.2.tar.gz) and [10.22.0](https://azcopyvnext.azureedge.net/releases/release-10.22.0-20231205/azcopy_linux_amd64_10.22.0.tar.gz), and it doesn't exactly "work", but I think I understand what @smalis-msft was thinking.

`AZCOPY_AUTO_LOGIN_TYPE=device ~/bin/azcopy_linux_amd64_10.21.2/azcopy ls https://***.blob.core.***/***/` Does work, while in 10.22.0, it fails

```
failed to traverse container: cannot list files due to reason GET https://***
--------------------------------------------------------------------------------
RESPONSE 401: 401 Server failed to authenticate the request. Please refer to the information in the www-authenticate header.
ERROR CODE: NoAuthenticationInformation
--------------------------------------------------------------------------------
﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>NoAuthenticationInformation</Code><Message>Server failed to authenticate the request. Please refer to the information in the www-authenticate header.
RequestId:***
Time:2024-12-***</Message></Error>
--------------------------------------------------------------------------------
```

So while the `ls` command "works" in this example, what it is _not_ doing is caching the credentials. The `_uid_ses.1000` keyring is still empty/not used. You have to log in every single time you run the `ls`. So the underlying issue of being unable to use the `_uid_ses.1000` keyring remains.

I'm guessing what happened between 10.22.0 and 10.21.2:
- 10.21.2 and older: Will get the session token, save it to the keystore, and then use the key it received
- 10.22.0 and newer: Will get the session token, save it to the keystore, load it from the keystore, and then use the key it loaded from the keystore

Since `azcopy` is unable to load _or_ save from `_uid_ses.1000`, `10.22.0` and newer will never work, and the older version can "work once per login"
